File: python/mirage/workspace/session/resolve.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Mapping

from mirage.policy.errors import PolicyError
from mirage.policy.types import AdmissionRules, CommandRule
from mirage.types import HiddenPaths, MountMode, weaker_mode
from mirage.utils.hidden import classify_paths, classify_vars
from mirage.workspace.session.constants import DEFAULT_PROFILE
from mirage.workspace.session.session import Session, vars_from_env
from mirage.workspace.session.shell_dirs import set_cwd

from mirage.workspace.session.permissions import (  # isort: skip
    CommandsBlock, CompiledProfile, MountCommandsBlock, PathsBlock,
    ProfileMount, SessionProfile, VarsBlock)


def resolve_profile(
    profiles: Mapping[str, SessionProfile],
    profile: str | SessionProfile | None,
) -> SessionProfile | None:
    """The role a session is created from.

    A name is looked up as written; a profile object is itself; None
    picks ``profiles.default`` when the workspace defines one and
    leaves the session unrestricted otherwise. There is no inheritance
    chain: a role is the whole document, so nothing is assembled from
    somewhere else before it is read.

    Args:
        profiles (Mapping[str, SessionProfile]): the workspace's named
            roles.
        profile (str | SessionProfile | None): what ``create_session``
            was given.

    Raises:
        PolicyError: the name is not a role the workspace defines.
    """
    if profile is None:
        return profiles.get(DEFAULT_PROFILE)
    if not isinstance(profile, str):
        return profile
    if profile not in profiles:
        raise PolicyError(f"unknown profile {profile!r}")
    return profiles[profile]


def _union_hide(a: PathsBlock | VarsBlock | None,
                b: PathsBlock | VarsBlock | None) -> tuple[str, ...]:
    """Every entry of both blocks, first spelling wins, order kept.

    Args:
        a (PathsBlock | VarsBlock | None): the role's block.
        b (PathsBlock | VarsBlock | None): the inline block.
    """
    out: list[str] = []
    for block in (a, b):
        for entry in (block.hide if block is not None else ()):
            if entry not in out:
                out.append(entry)
    return tuple(out)


def refuse_allow(inline: CommandsBlock | None) -> None:
    """Refuse an allow list in an inline document.

    The refusal belongs to *where the document was written*, not to
    whether a role happened to resolve, so both paths into
    ``with_inline`` run it: a workspace with no default role must not
    quietly accept a list a workspace with one refuses.

    Args:
        inline (CommandsBlock | None): what ``create_session`` added.

    Raises:
        PolicyError: the inline document states an allow list.
    """
    if inline is not None and inline.allow is not None:
        raise PolicyError("inline permissions may add ask and deny rules, "
                          "not an allow list")


def _add_commands(base: CommandsBlock | None,
                  inline: CommandsBlock | None) -> CommandsBlock | None:
    """The role's commands block with the inline document's rules added.

    An inline document may only restrict, so it carries ask and deny
    rules and never an allow list: a list there would install a command
    the role does not have, which is the one thing a per-call document
    must not do.

    Args:
        base (CommandsBlock | None): the role's block.
        inline (CommandsBlock | None): what ``create_session`` added.

    Raises:
        PolicyError: the inline document states an allow list.
    """
    if inline is None:
        return base
    refuse_allow(inline)
    if base is None:
        return inline
    return CommandsBlock(allow=base.allow,
                         ask=base.ask + inline.ask,
                         deny=base.deny + inline.deny)


def _add_mount(base: ProfileMount | None,
               inline: ProfileMount | None) -> ProfileMount:
    """One mount's entry with the inline document's added: the weaker
    mode, both rule lists, both hide lists.

    Args:
        base (ProfileMount | None): the role's entry, None when it
            names no settings for this mount.
        inline (ProfileMount | None): the inline entry.
    """
    if base is None:
        return inline if inline is not None else ProfileMount()
    if inline is None:
        return base
    mode = base.mode
    if inline.mode is not None:
        mode = (inline.mode if mode is None else weaker_mode(
            mode, inline.mode))
    ask = _rules_of(base.commands, "ask") + _rules_of(inline.commands, "ask")
    deny = _rules_of(base.commands, "deny") + _rules_of(
        inline.commands, "deny")
    hide = _union_hide(base.paths, inline.paths)
    return ProfileMount(mode=mode,
                        commands=(MountCommandsBlock(ask=ask, deny=deny) if
                                  (ask or deny) else None),
                        paths=PathsBlock(hide=hide) if hide else None)


def _rules_of(block: MountCommandsBlock | None,
              verb: str) -> tuple[CommandRule, ...]:
    """One verb's rules in a mount entry's commands block, empty when
    unstated.

    Args:
        block (MountCommandsBlock | None): the entry's block.
        verb (str): ``ask`` or ``deny``.
    """
    if block is None:
        return ()
    return block.ask if verb == "ask" else block.deny


def with_inline(base: SessionProfile | None,
                inline: SessionProfile | None) -> SessionProfile | None:
    """A role with the inline document of one ``create_session`` added.

    The one rule about combining two documents: an inline document may
    add ask and deny rules and hides, never an allow list, and that
    holds even when there is no role to add to. Modes take the weaker
    of the two, ``cwd`` and ``env`` are the inline document's when it
    states them (they are session presets, not permissions). Either
    side None returns the other unchanged.

    Args:
        base (SessionProfile | None): the resolved role.
        inline (SessionProfile | None): what ``create_session`` added.
    """
    if inline is None:
        return base
    refuse_allow(inline.commands)
    if base is None:
        return inline
    hide_paths = _union_hide(base.paths, inline.paths)
    hide_vars = _union_hide(base.vars, inline.vars)
    env = None
    if base.env is not None or inline.env is not None:
        env = {**(base.env or {}), **(inline.env or {})}
    mounts: dict[str, ProfileMount] | None = None
    if base.mounts is not None or inline.mounts is not None:
        prefixes = [*(base.mounts or {})]
        prefixes.extend(p for p in (inline.mounts or {}) if p not in prefixes)
        mounts = {
            prefix:
            _add_mount((base.mounts or {}).get(prefix), (inline.mounts
                                                         or {}).get(prefix))
            for prefix in prefixes
        }
    return SessionProfile(
        cwd=inline.cwd if inline.cwd is not None else base.cwd,
        env=env,
        mounts=mounts,
        paths=(PathsBlock(hide=hide_paths) if
               (base.paths is not None or inline.paths is not None) else None),
        vars=(VarsBlock(hide=hide_vars) if
              (base.vars is not None or inline.vars is not None) else None),
        commands=_add_commands(base.commands, inline.commands),
    )


def _root_of(prefix: str) -> str:
    """One spelling for a mount prefix: leading slash, no trailing one.

    Args:
        prefix (str): the prefix as the document spells it.
    """
    return "/" + prefix.strip("/")


def _anchored(entries: tuple[str, ...], root: str) -> tuple[str, ...]:
    """A mount section's path entries, anchored to the mount they are
    written under.

    An absolute entry already names something inside the root
    (``_under_mount`` refuses one that does not) and is left as written.
    A name pattern (``*.pem``, no slash) anchors nothing, and both
    places a mount section's entries are read from have lost the
    section by then: the session's hidden set is one list for every
    mount, and the op door matches a rule's paths without consulting
    ``rule.mount``. Left raw, ``mounts./repo.paths.hide: ["*.pem"]``
    hid ``/other/key.pem`` too, and a path-only deny under ``/repo``
    refused a read of it. The dialect's ``*`` crosses ``/``, so
    ``/repo/*.pem`` is every ``.pem`` at any depth below ``/repo`` and
    nothing outside it; anchoring also gives the entry the mount's own
    anchor depth, which is what it was always worth.

    Args:
        entries (tuple[str, ...]): the entries as written.
        root (str): the mount root, leading slash, no trailing one.
    """
    return tuple(e if e.startswith("/") else f"{root.rstrip('/')}/{e}"
                 for e in entries)


def _scoped_rules(rules: tuple[CommandRule, ...],
                  root: str) -> tuple[CommandRule, ...]:
    """A mount entry's rules, stamped with the mount they belong to and
    anchored to it.

    The stamp is what makes the rule apply to a line that *works
    inside* the mount, by cwd or by operand, which a path-scoped rule
    cannot express. The anchor is for the entries the stamp cannot
    reach: the op door reads a rule's paths alone (:func:`_anchored`).

    Args:
        rules (tuple[CommandRule, ...]): the rules as written.
        root (str): the mount prefix, leading slash, no trailing one.
    """
    return tuple(
        CommandRule(reason=rule.reason,
                    commands=rule.commands,
                    paths=_anchored(rule.paths, root),
                    mount=root) for rule in rules)


def compile_commands(profile: SessionProfile) -> AdmissionRules | None:
    """A role's admission rules: its own, plus every mount entry's,
    in one list; None when the role states none.

    Mount rules come first so the entry closest to the data speaks
    first when several rules match at the same anchor depth and only
    the message differs.

    Args:
        profile (SessionProfile): the resolved role.
    """
    ask: list[CommandRule] = []
    deny: list[CommandRule] = []
    for prefix, entry in (profile.mounts or {}).items():
        root = _root_of(prefix)
        ask.extend(_scoped_rules(_rules_of(entry.commands, "ask"), root))
        deny.extend(_scoped_rules(_rules_of(entry.commands, "deny"), root))
    block = profile.commands
    allow = block.allow if block is not None else None
    if block is not None:
        ask.extend(block.ask)
        deny.extend(block.deny)
    if allow is None and not ask and not deny:
        return None
    return AdmissionRules(allow=allow, ask=tuple(ask), deny=tuple(deny))


def _hidden(profile: SessionProfile) -> HiddenPaths | None:
    """Every path the role hides: its own entries, and each mount
    entry's anchored to the mount it was written under, since the set
    is one list for the whole session and nothing in it remembers which
    section an entry came from (:func:`_anchored`).

    Args:
        profile (SessionProfile): the resolved role.
    """
    entries: list[str] = []
    if profile.paths is not None:
        entries.extend(profile.paths.hide)
    for prefix, entry in (profile.mounts or {}).items():
        if entry.paths is not None:
            entries.extend(_anchored(entry.paths.hide, _root_of(prefix)))
    return classify_paths(entries)


def _modes(profile: SessionProfile) -> dict[str, MountMode] | None:
    """The mode each mount section states, None when none does.

    A mount the role does not name is absent from the map and keeps
    the mode it declares in the workspace's ``mounts:``; the map only
    narrows, it never grants.

    Args:
        profile (SessionProfile): the resolved role.
    """
    modes = {
        prefix: entry.mode
        for prefix, entry in (profile.mounts or {}).items()
        if entry.mode is not None
    }
    return modes or None


def compile_profile(effective: SessionProfile | None) -> CompiledProfile:
    """The session fields a role compiles to.

    Args:
        effective (SessionProfile | None): the resolved role with any
            inline document already added; None is an unrestricted
            session.
    """
    if effective is None:
        return CompiledProfile(mount_modes=None,
                               hidden_paths=None,
                               hidden_vars=None,
                               env=None,
                               cwd=None,
                               commands=None)
    return CompiledProfile(
        mount_modes=_modes(effective),
        hidden_paths=_hidden(effective),
        hidden_vars=classify_vars(
            effective.vars.hide if effective.vars is not None else ()),
        env=dict(effective.env) if effective.env is not None else None,
        cwd=effective.cwd,
        commands=compile_commands(effective),
    )


def narrow(session: Session, compiled: CompiledProfile) -> None:
    """Stamp a compiled role's narrowing onto a session.

    The four fields no shell line can edit: the per-mount modes, hidden
    paths, hidden variables, the admission rules. Applied at creation
    and again whenever a stored record could carry a stale copy (the
    default session after hydration), so the document, not the store,
    is what an agent runs under.

    Args:
        session (Session): the session to narrow.
        compiled (CompiledProfile): the effective role.
    """
    session.mount_modes = (dict(compiled.mount_modes)
                           if compiled.mount_modes is not None else None)
    session.hidden_paths = compiled.hidden_paths
    session.hidden_vars = compiled.hidden_vars
    session.commands = compiled.commands


def apply_profile(session: Session, compiled: CompiledProfile) -> None:
    """Narrow a fresh session and seed its scratch state from the role.

    A role's env is a *process* environment, the same shape
    ``ws.env = {...}`` speaks, so every name in it is exported: seeding
    them plain left ``$TOKEN`` expanding while every command, CLI and
    guest runtime in the profiled session saw nothing, since all three
    read ``env_snapshot`` and that is the exported set. The cwd is where
    the session starts; both are the agent's to change afterwards,
    which is why hydration keeps the stored ones and re-stamps only
    :func:`narrow`.

    Args:
        session (Session): the session just created.
        compiled (CompiledProfile): the effective role.
    """
    narrow(session, compiled)
    if compiled.env:
        session.vars.update(vars_from_env(compiled.env))
    if compiled.cwd is not None:
        set_cwd(session, compiled.cwd)

File: typescript/packages/core/src/workspace/session/resolve.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { PolicyError } from '../../policy/errors.ts'
import type { HiddenPaths } from '../../types.ts'
import { MountMode, weakerMode } from '../../types.ts'
import { classifyPaths, classifyVars } from '../../utils/hidden.ts'
import { stripSlash } from '../../utils/slash.ts'
import {
  parseProfileMounts,
  type CompiledProfile,
  type MountPermissions,
  type PathsBlock,
  type SessionProfile,
  type VarsBlock,
  type WorkspacePermissions,
} from './profile.ts'

export const DEFAULT_PROFILE = 'default'

/** A profile's mounts once every spelling is normalized: ceilings, an allowlist, or unrestricted. */
export type ProfileMounts = ReadonlyMap<string, MountMode> | readonly string[] | null

/**
 * Resolve a named profile through its `extends` chain.
 *
 * Field inheritance, root first: a stated field replaces the parent's,
 * an absent one is inherited. Safety comes from the layer intersection
 * at evaluation, not from inheritance, so a child may state fewer hides
 * than its parent. The result names no parent. Throws PolicyError on an
 * unknown name or a cycle.
 */
export function inherit(
  profiles: Readonly<Record<string, SessionProfile>>,
  name: string,
): SessionProfile {
  const chain: SessionProfile[] = []
  const seen: string[] = []
  let current: string | null | undefined = name
  while (current !== null && current !== undefined) {
    if (seen.includes(current)) {
      throw new PolicyError(`profile extends cycle: ${[...seen, current].join(' -> ')}`)
    }
    const node: SessionProfile | undefined = Object.prototype.hasOwnProperty.call(profiles, current)
      ? profiles[current]
      : undefined
    if (node === undefined) {
      const parent = seen.at(-1)
      const where =
        parent !== undefined ? `profile '${parent}' extends unknown profile` : 'unknown profile'
      throw new PolicyError(`${where} '${current}'`)
    }
    seen.push(current)
    chain.push(node)
    current = node.extends
  }
  const merged: {
    cwd?: string | null
    env?: Readonly<Record<string, string>> | null
    mounts?:
      | ReadonlyMap<string, string>
      | Readonly<Record<string, string>>
      | readonly string[]
      | null
    paths?: PathsBlock | null
    vars?: VarsBlock | null
  } = {}
  for (const node of [...chain].reverse()) {
    if (node.cwd != null) merged.cwd = node.cwd
    if (node.env != null) merged.env = node.env
    if (node.mounts != null) merged.mounts = node.mounts
    if (node.paths != null) merged.paths = node.paths
    if (node.vars != null) merged.vars = node.vars
  }
  return merged
}

/**
 * The profile a session is created from, before inline tightening: a
 * name resolves through `inherit`; a profile object that names a
 * parent resolves the same way with itself as the child; null picks
 * `profiles.default` when the workspace defines one and leaves the
 * session unrestricted otherwise.
 */
export function resolveProfile(
  profiles: Readonly<Record<string, SessionProfile>>,
  profile: string | SessionProfile | null | undefined,
): SessionProfile | null {
  if (profile === null || profile === undefined) {
    return Object.prototype.hasOwnProperty.call(profiles, DEFAULT_PROFILE)
      ? inherit(profiles, DEFAULT_PROFILE)
      : null
  }
  if (typeof profile === 'string') return inherit(profiles, profile)
  if (profile.extends == null) return profile
  return inherit({ ...profiles, '': profile }, '')
}

/**
 * The mounts both sides grant, at the weaker mode. A Map is a set of
 * ceilings, an array an allowlist at each mount's own mode. Map x Map:
 * common prefixes at the weaker mode; Map x array: the Map's entries the
 * list also names; array x array: the common prefixes; null x anything:
 * the other.
 */
function isCeilings(mounts: ProfileMounts): mounts is ReadonlyMap<string, MountMode> {
  return mounts instanceof Map
}

function intersectMounts(base: ProfileMounts, inline: ProfileMounts): ProfileMounts {
  if (base === null) return inline
  if (inline === null) return base
  const baseCeilings = isCeilings(base) ? base : null
  const inlineCeilings = isCeilings(inline) ? inline : null
  const baseListed = isCeilings(base) ? null : base
  const inlineListed = isCeilings(inline) ? null : inline
  if (baseCeilings !== null && inlineCeilings !== null) {
    const out = new Map<string, MountMode>()
    for (const [p, m] of baseCeilings) {
      const other = inlineCeilings.get(p)
      if (other !== undefined) out.set(p, weakerMode(m, other))
    }
    return out
  }
  if (baseCeilings !== null && inlineListed !== null) {
    return new Map([...baseCeilings].filter(([p]) => inlineListed.includes(p)))
  }
  if (inlineCeilings !== null && baseListed !== null) {
    return new Map([...inlineCeilings].filter(([p]) => baseListed.includes(p)))
  }
  if (baseListed !== null && inlineListed !== null)
    return baseListed.filter((p) => inlineListed.includes(p))
  return null
}

/** Every entry of both blocks, first spelling wins, order kept. */
function unionHide(
  a: PathsBlock | VarsBlock | null | undefined,
  b: PathsBlock | VarsBlock | null | undefined,
): string[] {
  const out: string[] = []
  for (const block of [a, b]) {
    for (const entry of block?.hide ?? []) if (!out.includes(entry)) out.push(entry)
  }
  return out
}

/**
 * Narrow a profile by an inline document (design 3.4): mounts
 * intersect, hides union, `cwd` and `env` are the inline document's
 * when it states them (session presets, not permissions). Either side
 * null returns the other unchanged.
 */
export function tighten(
  base: SessionProfile | null,
  inline: SessionProfile | null,
): SessionProfile | null {
  if (base === null) return inline
  if (inline === null) return base
  const out: {
    cwd?: string | null
    env?: Readonly<Record<string, string>> | null
    mounts?: ProfileMounts
    paths?: PathsBlock | null
    vars?: VarsBlock | null
  } = {}
  const cwd = inline.cwd ?? base.cwd
  if (cwd != null) out.cwd = cwd
  if (base.env != null || inline.env != null)
    out.env = { ...(base.env ?? {}), ...(inline.env ?? {}) }
  const mounts = intersectMounts(parseProfileMounts(base.mounts), parseProfileMounts(inline.mounts))
  if (mounts !== null) out.mounts = mounts
  if (base.paths != null || inline.paths != null)
    out.paths = { hide: unionHide(base.paths, inline.paths) }
  if (base.vars != null || inline.vars != null)
    out.vars = { hide: unionHide(base.vars, inline.vars) }
  return out
}

/**
 * A mount's hides in absolute terms. Every entry, glob or plain, is
 * joined under the mount root, so a mount-relative rule can never reach
 * outside its mount; a slashless glob stops being a component pattern
 * and becomes anchored under the mount, which is the only reading
 * "relative to the mount root" can have.
 */
export function rebase(prefix: string, perms: MountPermissions | null | undefined): string[] {
  if (perms == null) return []
  const root = '/' + stripSlash(prefix)
  const base = root === '/' ? '' : root
  return perms.paths.hide.map((entry) => {
    const rel = entry.replace(/^\/+/, '')
    return rel === '' ? root : `${base}/${rel}`
  })
}

/**
 * What every session of the workspace cannot see: the workspace tier's
 * hides plus each mount's rebased hides, compiled once and stamped onto
 * every session by the session manager, joined with the session's own
 * hides in the predicate.
 */
export function boundHidden(
  workspace: WorkspacePermissions | null | undefined,
  mounts: ReadonlyMap<string, MountPermissions | null | undefined>,
): HiddenPaths | null {
  const entries: string[] = []
  if (workspace != null) entries.push(...workspace.paths.hide)
  for (const [prefix, perms] of mounts) entries.push(...rebase(prefix, perms))
  return classifyPaths(entries)
}

/** The session fields an effective profile sets; null is an unrestricted session. */
export function compileProfile(effective: SessionProfile | null): CompiledProfile {
  if (effective === null) {
    return { mountModes: null, hiddenPaths: null, hiddenVars: null, env: null, cwd: null }
  }
  const mounts = parseProfileMounts(effective.mounts)
  let modes: ReadonlyMap<string, MountMode> | null = null
  if (isCeilings(mounts)) modes = mounts
  else if (mounts !== null)
    modes = new Map(mounts.map((p): [string, MountMode] => [p, MountMode.EXEC]))
  return {
    mountModes: modes,
    hiddenPaths: classifyPaths(effective.paths?.hide ?? []),
    hiddenVars: classifyVars(effective.vars?.hide ?? []),
    env: effective.env ?? null,
    cwd: effective.cwd ?? null,
  }
}
